jobqueue_features/clusters.py

```diff
@@ -614,77 +744,57 @@ def __del__(self):
+    __doc__ = f"""Custom PBS Cluster class.
 
-    def _add_to_cluster_controller(self):
-        from .clusters_controller import clusters_controller_singleton
-
-        clusters_controller_singleton.add_cluster(id_=self.name, cluster=self)
+    Attributes:
+    ---------
+    ngpus_per_node : int
+        The number of gpus per node to be used
+    {custom_cluster_attributes}
+    """
 
-    def _update_script_nodes(self, **kwargs) -> None:
-        # If we're not in mpi_mode no need to do anything
-        if not self.mpi_mode:
-            return
+    job_cls = CustomPBSJob
 
-        # When in MPI mode, after jobqueue has initialised we update the jobscript with
-        # the `real` number of MPI tasks
-        self._kwargs["mpi_tasks"] = self.mpi_tasks
-
-        # The default for jobqueue is not to use an MPI launcher (since it is not MPI
-        # aware). However, if self.fork_mpi=False then the tasks intended for this
-        # cluster are MPI-enabled. In order to give them an MPI environment we need to
-        # use our custom wrapper and launch with our MPI launcher
-        if not self.fork_mpi:
-            command_template = self._dummy_job._command_template
-            dask_worker_module = "distributed.cli.dask_worker"
-            if dask_worker_module in command_template:
-                command_template = command_template.replace(
-                    dask_worker_module, MPI_DASK_WRAPPER_MODULE
-                )
-            else:
-                raise RuntimeError(
-                    "Python module {} not found in command template:\n{}".format(
-                        dask_worker_module, command_template
-                    )
+    def __init__(self, **kwargs):
+        self.scheduler_name = "pbs"
+        self.ngpus_per_node = kwargs.get("ngpus_per_node", 0)
+        kwargs = self.update_init_kwargs(**kwargs)
+        self.validate_cluster_name(kwargs["name"])
+        self.name = kwargs["name"]
+        if self.mpi_mode:
+            if "job_cpu" in kwargs:
+                raise ValueError(
+                    "We don't allow 'job_cpu' as a kwarg in MPI mode since we"
+                    " leverage this in jobqueue to set cpus-per-task"
                 )
-            # The first part of the string is the python executable to use for the
-            # worker
-            python, arguments = command_template.split(" ", 1)
-
-            # Wrap the launch command with our mpi wrapper
+            mpi_job_extra = []
+            if "job_extra" not in kwargs:
+                raise KeyError("job_extra keyword should always be set in kwargs")
+            mpi_job_extra.extend(kwargs["job_extra"])
+            kwargs.update({"job_extra": mpi_job_extra})
+        super().__init__(**kwargs)
+        self._update_script_nodes(**kwargs)
+        if self.mpi_mode:
+            if hasattr(self, "mpi_tasks"):
+                self._kwargs["mpi_tasks"] = self.mpi_tasks
+            if self.ngpus_per_node > 0:
```

Comment:
Actually, the GPUs you might want to request outside of MPI mode.

Reply:
Changed it.